### ISSUE TITLE
Native Kubernetes Deployment + Keycloak as OIDC instead of DEX - avoid looping

### ISSUE THREAD

**Issue by abdennour** [openshift/console#9754]
I tried to deploy the OKD origin-console using [this helm chart](https://artifacthub.io/packages/helm/av1o-charts/openshift-console). & using keycloak realm instead of DEX , on top of a native kubernetes v1.19.13

After installing the chart, my deployment looks like: 
```yaml
      - env:
        - name: BRIDGE_DOCUMENTATION_BASE_URL
          value: https://kubernetes.io/docs/
        - name: BRIDGE_DEX_API_HOST
          value: https://keycloak.example.co/auth/realms/realm-x
        - name: BRIDGE_BASE_ADDRESS
          value: https://console.devops.example.co
        - name: BRIDGE_USER_AUTH
          value: oidc
        - name: BRIDGE_K8S_AUTH
          value: oidc
        - name: BRIDGE_USER_AUTH_OIDC_ISSUER_URL
          value: https://keycloak.example.co/auth/realms/realm-x
        - name: BRIDGE_USER_AUTH_OIDC_CLIENT_ID
          value: console
        - name: BRIDGE_USER_AUTH_OIDC_CLIENT_SECRET
          value: yyyyyyy-xxxx-zzzz-tttt-xxxxxxxxxx
        image: quay.io/openshift/origin-console:4.6.0
```

The console is able to redirect me to the keycloak login page. However, after a success login, i cannot see everything in the console. I just see infinite loop of  page refresh.

These are my questions :
- Is it possible to integrate correctly with Keycloak as OIDC provider instead of Dex ?
- What are the missing env vars ? because after some research, i found [someone](https://www.reddit.com/r/kubernetes/comments/d50oe9/openshift_web_ui_on_upstream_k8s/f0jg8tl/) to refer to other env vars like : `BRIDGE_KUBECTL_CLIENT_ID`, `BRIDGE_KUBECTL_CLIENT_SECRET` which i don't know how to assign a value for them ?
- What are the required clusterrolebinding ?
- For the websocket, what's the required annotations at ingress level, so i can avoid failed websocket ?
- What are the CRDs that i must deploy to the native kubernetes ? so i can take the most benefit from the console . For example, the helm chart mentioned at the beginning deploys the CRD "app" of Openshift.

Note that [this configuration](https://github.com/openshift/console/issues/1377#issuecomment-604146607) is working with me. However, it does not delegate to users with RBAC.

**Comment by spadgett** [comment]
I have not tested running console with Keycloak unfortunately. Are there any errors in the console pod log that might indicate what's wrong?

The only config that looks like it might be missing is `BRIDGE_USER_AUTH_OIDC_CA_FILE`. You should check if there are any certificate errors in the pod logs.

I don't believe `BRIDGE_DEX_API_HOST` is used by console.

I don't know offhand what ingress annotations you need for websockets. What ingress controller are you using?

@pedjak might be able to answer your question on the Helm CRD.

**Comment by LarsBingBong** [comment]
@abdennour where you able to get this to work? Thanks

**Comment by Ankitcapco** [comment]
HI All,

I am using following way to deploy OKD, I am using Microsoft Authenticator and able to successfully login  and in pod logs as well showing outh success. But after successfully login, it is going to BASE-URL and there it is going in an infinite loop. Is there any workaround for it? Please help,

![Screen Shot 2021-11-29 at 10 42 45 AM](https://user-images.githubusercontent.com/92543508/143898348-5e3bb169-0faf-4580-b3c7-5a071ad1eb5f.png)

            - name: BRIDGE_USER_AUTH_OIDC_ISSUER_URL
              value: https://sts.windows.net/********************/
            - name: BRIDGE_USER_AUTH_OIDC_CLIENT_ID
              value: *******************************
            - name: BRIDGE_USER_AUTH_OIDC_CLIENT_SECRET
              value: *******************************
            - name: BRIDGE_USER_AUTH
              value: oidc
            - name: BRIDGE_BASE_ADDRESS
              value: "https://console.internal.mgmt.platformzero.build"
            - name: BRIDGE_K8S_MODE
              value: in-cluster
            - name: BRIDGE_K8S_MODE_OFF_CLUSTER_ENDPOINT
              value: https://kubernetes.default #master api
            - name: BRIDGE_K8S_MODE_OFF_CLUSTER_SKIP_VERIFY_TLS
              value: "true" # no tls enabled
            - name: BRIDGE_K8S_AUTH
              value: bearer-token
            - name: BRIDGE_K8S_AUTH_BEARER_TOKEN
              valueFrom:
                secretKeyRef:
                  name: console-token-##### console serviceaccount token
                  key: token

**Comment by Ankitcapco** [comment]
hi @abdennour any chance, could you solve this issue "infinite loop of page refresh". I am getting stuck at same place. After successful login, it is just refreshing page again and again

**Comment by abdennour** [comment]
Hi @Ankitcapco , i didn't get a resolution till now. I just ran the console on read mode without authentication integration.

**Comment by chris93111** [comment]
Hi any update ?

**Comment by Nurlan199206** [comment]
i have same issue with looping and websockets. console version: 4.14

**Comment by Cenness** [comment]
> Since this is the _OpenShift_ Console it works off the expectation that the Kubernetes APIServer will accept an OIDC token, which for most distributions it won't (at least out-of-the-box). When the OIDC token is not accepted, the Console goes into a reload loop which is extremely annoying and hard to debug.


I didn't figure that out without grokking the source, so I'll rephrase it in a more direct way for others:
You'll need to use the **same** oidc client for both your cluster and console.
You can use different `BRIDGE_USER_AUTH_OIDC_CLIENT_ID` and `BRIDGE_KUBECTL_CLIENT_ID` only if you using Dex, [this "hack" with audiences](https://github.com/openshift/console/blob/bfb8b47fc4cdeb5d9171fdc1a46dbcc99061b113/cmd/bridge/main.go#L667C3-L677) doesn't work with Keycloak.

**Comment by MadEngineX** [comment]
Faced this issue on our GKE cluster. 
1) We've enabled [anthos-indentity-service](https://cloud.google.com/kubernetes-engine/docs/how-to/oidc#enable-oidc) in GKE.
2) We configured client "kubernetes" in our Keycloak
3) Created k8s clusterRoleBinding: 
```yaml 
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  creationTimestamp: "2023-10-16T19:37:34Z"
  name: oidc-cluster-admin
  resourceVersion: "66912454"
  uid: 0d258571-db18-4786-8217-a4ff63b88fe2
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cluster-admin
subjects:
- apiGroup: rbac.authorization.k8s.io
  kind: Group
  name: oidc:admin
```
4) Added Mapper in Keycloak, like it described in kube-oidc-ligin docs: 
https://github.com/int128/kubelogin/blob/master/docs/setup.md#keycloak
5) Deployed openshift-console and faces ininitive loops. 
6) We went inside source code, and understood that our requests went to api-server directly againsts going to the anthos-indentity-service. So we need to add a few variables: 

- BRIDGE_K8S_MODE_OFF_CLUSTER_ENDPOINT=https://10.11.12.103:443 <- IP addres of service type LoadBalancer, assigned to our anthos-indentity-service.
- BRIDGE_K8S_MODE=off-cluster <- to let openshisft-console use anthos-indentity-service IP against k8s service of kube-api-server
- BRIDGE_K8S_MODE_OFF_CLUSTER_SKIP_VERIFY_TLS=true <- if you don't want to manage certs

Now everything is fine, we can add any Role in k8s and map it with Keycloak Client Role.

**Comment by cloudcafetech** [comment]
@djcass44 
I there any way to skip TLS for DEX.

I am getting following error.

```
E1002 04:04:04.254699       1 auth.go:232] error contacting auth provider (retrying in 10s): Get "https://auth.172.30.2.2.nip.io:30443/.well-known/openid-configuration": x509: certificate signed by unknown authority
E1002 04:04:14.268506       1 auth.go:232] error contacting auth provider (retrying in 10s): Get "https://auth.172.30.2.2.nip.io:30443/.well-known/openid-configuration": x509: certificate signed by unknown authority
E1002 04:04:25.358494       1 auth.go:232] error contacting auth provider (retrying in 10s): Get "https://auth.172.30.2.2.nip.io:30443/.well-known/openid-configuration": x509: certificate signed by unknown authority
```

**Comment by djcass44** [comment]
@cloudcafetech it doesn't look like it. You'll need to provide a CA using either the `user-auth-oidc-ca-file` flag, or by overriding the system trust with the `SSL_CERT_FILE` environment variable

**Comment by djcass44** [comment]
The OpenShift Console does not use the Kubernetes Metrics API, it calls
Prometheus directly. It has a hard coded URL but if you want to put in the
effort you can make it work. I've done it before, but IMO it's not worth
the effort.

**Comment by cloudcafetech** [comment]
@djcass44 

Thanks for quick reply.
If not metric server,  can we integrate with existing Prometheus ? please refer some links.

**Comment by gecube** [comment]
@spamcop Hi! We strongly believe that in cluster mode now is not very useful.
My good colleague prepared a gist and published it here: https://gist.github.com/identw/c9fbdd0fc22c0b766be7a794c3910b97
Please use.

**Comment by cloudcafetech** [comment]
@gecube using above gist getting following error.

**Fatal glibc error: CPU does not support x86-64-v2**

:(

**Comment by gecube** [comment]
@cloudcafetech Hi! It worked for me, probably you have an old OS, IDK. My base operating system for the cluster is bottlerocket. Also it works well on talos.

**Comment by cloudcafetech** [comment]
@gecube 
Thank you for data point, seem issue is old Hw and Proxmox CPU settings but it works with image version 4.10 with **observe** menu but facing below metrics which are showing **no data point found** 

![image](https://github.com/user-attachments/assets/83f2bece-f64c-4bc9-942c-72d6e7532ccb)


```
sum(kube_pod_resource_request{node='mgm-master1',resource='cpu'})

sum(kube_pod_resource_limit{node='mgm-master1',resource='cpu'})

instance:node_cpu:rate:sum{instance='mgm-master1'}

instance:node_network_transmit_bytes:rate:sum{instance='mgm-master1'}

instance:node_network_receive_bytes:rate:sum{instance='mgm-master1'}

kubelet_running_pods{instance=~'192.168.0.111:.*'}
```

**Comment by gecube** [comment]
@cloudcafetech Hi! I may guess that you have incorrect settings for integration with prometheus. It is difficult to say as I am not developer and from my POV openshift-console is a little bit bloated and designed like a blackbox, very tightly coupled to Openshift distribution, so it is not so easy to adapt it to vanilla k8s. With all respect to engineers who built it. I think we can find out what are the real dependencies and which settings you need to apply. I think that you need to properly set up 

```
BRIDGE_ALERTMANAGER_PUBLIC_URL
BRIDGE_THANOS_PUBLIC_URL
```

like here : https://paste.openstack.org/show/bFzlEFaoPgSLFUHKZxUs/ or even use 

```
BRIDGE_K8S_MODE_OFF_CLUSTER_ALERTMANAGER
BRIDGE_K8S_MODE_OFF_CLUSTER_THANOS
```

I want to emphasise that I am not openshift-console specialist, so I am also forced to find out with different approaches the proper settings.

**Comment by abdennour** [comment]
> 2\. configured



@MadEngineX can you share full-way how u integrated with gke ?

**Comment by MadEngineX** [comment]
> [@MadEngineX](https://github.com/MadEngineX) can you share full-way how u integrated with gke ?

@abdennour 

https://github.com/openshift/console/issues/9754#issuecomment-2071734589

In steps 1-4, I enabled the GKE Identity Service and linked it to my company's Keycloak. This allowed me to use OIDC login via kubectl.
But using kubectl is difficult for regular developers. To make access easier, I deployed this OKD Console:

```
helm repo add av1o https://av1o.gitlab.io/charts
helm upgrade --install console -f values.yaml av1o/openshift-console --namespace infra-openshift-console --create-namespace 
```
[values.txt](https://github.com/user-attachments/files/19254698/values.txt)



So I can assign Keycloak roles to my company's developers and they have access, limited by linked K8s RBAC roles.
